### ISSUE TITLE
Does WSL2 corrupted my git repository?

### ISSUE THREAD

**Issue by skhaz** [microsoft/WSL2-Linux-Kernel#168]
I recently switched my development enviroment from macOS to Windows and I am using WSL2 on my machine with VSCode and WSL2 remote extension.

Everthing was fine when today I got this error message from git

```
$ git pull
error: object file .git/objects/f8/72e0c57726ea0b5456269eebd81dd78f625980 is empty
fatal: loose object f872e0c57726ea0b5456269eebd81dd78f625980 (stored in .git/objects/f8/72e0c57726ea0b5456269eebd81dd78f625980) is corrupt
```

```
$ git status
error: object file .git/objects/f8/72e0c57726ea0b5456269eebd81dd78f625980 is empty
error: object file .git/objects/f8/72e0c57726ea0b5456269eebd81dd78f625980 is empty
error: object file .git/objects/f8/72e0c57726ea0b5456269eebd81dd78f625980 is empty
fatal: loose object f872e0c57726ea0b5456269eebd81dd78f625980 (stored in .git/objects/f8/72e0c57726ea0b5456269eebd81dd78f625980) is corrupt
```

Any clues?

**Comment by microhobby** [comment]
@skhaz are you using through the Linux file system or the Windows file system?

**Comment by skhaz** [comment]
@microhobby through the Linux file system. All my code, git, tools and compilers are on Linux side.

**Comment by 2n3906** [comment]
This exact thing happened to me.  I'm also running WSL2 with Git on the Linux filesystem using VS Code to edit.

**Comment by PulsarFox** [comment]
I'm not on a laptop, and I have this error message everytime on a git repo, did a `find .git/objects/ -size 0 -delete` But then I had theses messages 
```
error: refs/heads/develop does not point to a valid object!
error: refs/remotes/origin/develop does not point to a valid object!
```
I pulled the project and now everything is fine, I didn't push yet so i hope it wont corrupt my remote branch too

**Comment by skhaz** [comment]
Just out of curiosity, are you guys using VSCode's terminal or something else?

**Comment by PulsarFox** [comment]
VSCode Integrated wsl terminal yes. Opened the project with the command `code .` inside the project folder

**Comment by 2n3906** [comment]
I was using VS Code inside WSL2 to edit files, but doing git directly from the latest Windows Terminal.

**Comment by skhaz** [comment]
Hmm... I have a suspicious, can be VSCode's fault, not sure, never happened to me again.

**Comment by Annih** [comment]
Hum it happens to me multiple times recently, and I'm not using VSCode.
It seems that it happens when I'm a bit too quick to close my WSL session (sometimes with a wsl shutdown) but not only.

**Comment by Champkinz** [comment]
Is there any direct solution for this, this happens once a day for me!

**Comment by sidharthramesh** [comment]
Happens to me too

**Comment by skhaz** [comment]
@sidharthramesh @Champkinz try to disable fast boot, never happened again.

**Comment by edrevo** [comment]
Related issues:
https://github.com/microsoft/WSL/issues/5026
https://github.com/microsoft/WSL/issues/5895

**Comment by Champkinz** [comment]
@skhaz So far no issues, but I managed reproduce the problem once, if I shutdown the PC while VSCODE is open , the git gets corrupted, but I tried like 3 times, only once the git got corrupted. just leaving this here for future reference.

**Comment by mbrumlow** [comment]
Just want to chime in, switching form a full Linux setup (actually installed windows for the first time in 15 years). And today I had my first corrupt git repo.

I use X with Emacs, and the command line only. No fancy external tools.  The error indicated a object file simply went missing. 

I suspect that some sort of optimization is keeping the fs from flushing. As the night before I had committed, and pushed, then installed a windows update that required a reboot. Today I continued working and wen i went to commit produced the object missing error.

**Comment by Champkinz** [comment]
After disabling fast boot, my issue was completely fixed.

**Comment by luigimannoni** [comment]
Disabling fast boot didn't help in my case, corruptions are more or less constant with fast boot switched on or off.

**Comment by skhaz** [comment]
It happened again, but this time with bash history got corrupted.

**Comment by Champkinz** [comment]
My bash history keeps getting corrupted every now and then, but not my git. 

Yesterday my whole npm package got misplaced or something , had to fix the path and reinstall it.

Feel like booting Linux on a separate disk because of these issues.

This is why mac is convenient when it comes to development.

**Comment by jmfury** [comment]
I have disabled fast boot and this the `.git` issue (`error: object file ... is empty` & `fatal: loose object ... is corrupt`) is still occurring

**Comment by JasonYangShadow** [comment]
WSL2 uses a hypervisor machine?  this issue might be caused by the virtual disk?

**Comment by QEDK** [comment]
Fwiw, I use WSL2 (latest Debian and Kali) without VSCode for ~6 months and this hasn't happened to me once - maybe because I never shutdown my laptop? 🤔

**Comment by dscho** [comment]
Have you tried fiddling with [`core.fsyncObjectFiles`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-corefsyncObjectFiles)?

**Comment by runvnc** [comment]
> Is there any direct solution for this, this happens once a day for me!

Use actual Linux such as Debian, Ubuntu or Pop OS.  If necessary, you can run Windows inside VirtualBox.  Or run Windows programs with Wine, Proton etc.  And it doesn't corrupt your file system.

**Comment by Champkinz** [comment]
> > Is there any direct solution for this, this happens once a day for me!
> 
> Use actual Linux such as Debian, Ubuntu or Pop OS.  If necessary, you can run Windows inside VirtualBox.  Or run Windows programs with Wine, Proton etc.  And it doesn't corrupt your file system.

This is gonna be my solution

**Comment by luigimannoni** [comment]
> > Is there any direct solution for this, this happens once a day for me!
> 
> Use actual Linux such as Debian, Ubuntu or Pop OS. If necessary, you can run Windows inside VirtualBox. Or run Windows programs with Wine, Proton etc. And it doesn't corrupt your file system.

While I would do this for a tower/desktop that isn't the solution for a laptop.

Mind that I come from a Ubuntu distro, Windows is the most energy efficient solution at the moment and allows me to run my Dev tools and work on battery for 4 hours straight. Ubuntu (or any Debian-based distro) destroy my battery in 40 minutes and there's no solution or optimization for that.

**Comment by timvisee** [comment]
I don't think there is space here for a Windows versus Linux discussion. This is something that should be fixed, likely on the Windows side of things. The cause of this issue might be rooted in Hyper-V's mounts.

**Comment by anaisbetts** [comment]
This is a dupe of https://github.com/microsoft/WSL/issues/5026, you can work around it by making sure not to use `shutdown /r /t 0` to shut the machine down.

**Comment by QEDK** [comment]
> 
> 
> > Is there any direct solution for this, this happens once a day for me!
> 
> Use actual Linux such as Debian, Ubuntu or Pop OS. If necessary, you can run Windows inside VirtualBox. Or run Windows programs with Wine, Proton etc. And it doesn't corrupt your file system.

This is not a solution for people who intend to use WSL2. And chances are people who don't want to use it already use something else.

**Comment by luigimannoni** [comment]
> This is a dupe of [microsoft/WSL#5026](https://github.com/microsoft/WSL/issues/5026), you can work around it by making sure not to use `shutdown /r /t 0` to shut the machine down. WSL isn't getting shut down properly on your computer and you're doing the equivalent of yanking the power on the device

Unfortunately that is not the case to many, this happens personally to me pretty much by shutting down the computer by using the shutdown button in the start menu. The problem appears also while entering and exiting standby and rebooting through windows update

**Comment by jeremyBanks** [comment]
This happens frequently if I accidentally interact with the same files from both the Windows and Linux side in close succession. I never explicitly shut down WSL. Some files seem to get stuck in a locked state that doesn't clear until I reboot the machine, while other files are not, and the result is an inconsistent/corrupt repo or project.

**Comment by ankushnarula** [comment]
For those experiencing this issue, have you tried testing using the `sync` and/or `halt` commands before exiting WSL2?

**Comment by timvisee** [comment]
@ankushnarula Some suggested `shutdown`, apparently with no effect. That implicitly does the same thing, syncing the underlying file system.

**Comment by ankushnarula** [comment]
> @ankushnarula Some suggested `shutdown`, apparently with no effect. That implicitly does the same thing, syncing the underlying file system.

Yes - implicitly.  Which is why I asked if anyone tried it explicitly. ;-)

**Comment by riverar** [comment]
> This is a dupe of [microsoft/WSL#5026](https://github.com/microsoft/WSL/issues/5026), you can work around it by making sure not to use `shutdown /r /t 0` to shut the machine down. WSL isn't getting shut down properly on your computer and you're doing the equivalent of yanking the power on the device

`shutdown /r /t 0` just triggers a normal shutdown and reboot, this is not an accurate description of what's going on at all.

**Comment by anaisbetts** [comment]
@riverar yes, on the Windows side. But the WSL VM isn't being shut down cleanly in this scenario - this is a bug

**Comment by ElTimablo** [comment]
> > > Is there any direct solution for this, this happens once a day for me!
> > 
> > 
> > Use actual Linux such as Debian, Ubuntu or Pop OS. If necessary, you can run Windows inside VirtualBox. Or run Windows programs with Wine, Proton etc. And it doesn't corrupt your file system.
> 
> While I would do this for a tower/desktop that isn't an helpful solution for a laptop.
> 
> Mind that I come from a Ubuntu distro, Windows is the most energy efficient solution at the moment and allows me to run my Dev tools and work on battery for 4 hours straight. Ubuntu (or any Debian-based distro) destroy my battery in 40 minutes and there's no solution or optimization for that.

There is a solution for that: turn off your dedicated GPU, because that's the only possible way I can see that happening on Ubuntu. Going around crowing about the lack of a solution just shows how little you actually care about fixing it.

**Comment by riverar** [comment]
@anaisbetts If that's true, it would also shutdown incorrectly during servicing, Start > Power > Restart, and app install. Or are you saying that's the bug? If so, that's a pretty serious oversight!

**Comment by fabianhjr** [comment]
> Ubuntu (or any Debian-based distro) destroy my battery in 40 minutes and there's no solution or optimization for that.

You could use a different governor from the `performance` default; if you want maximum battery time then go for the `powersave` governor though `ondemand` or `conservative` should give you better battery time without a significant performance penalty. ( https://wiki.archlinux.org/index.php/CPU_frequency_scaling / https://wiki.gentoo.org/wiki/Power_management/Processor )

**Comment by cmurf** [comment]
Folks having this problem could search the systemd journal for:
```
systemd-fsck[681]: /dev/vdb2: recovering journal
EXT4-fs (vdb2): recovery complete
```
Either being present supports the hypothesis that the file system is not being cleanly unmounted. Or possibly the host side cache isn't being flushed to stable media.

**Comment by piscisaureus** [comment]
This has happened to me in WSL1 too when I moved a large repo from `/mnt/c/some/repo` to `~/repo` using Linux `mv`.

**Comment by luigimannoni** [comment]
> > > > Is there any direct solution for this, this happens once a day for me!
> > > 
> > > 
> > > Use actual Linux such as Debian, Ubuntu or Pop OS. If necessary, you can run Windows inside VirtualBox. Or run Windows programs with Wine, Proton etc. And it doesn't corrupt your file system.
> > 
> > 
> > While I would do this for a tower/desktop that isn't an helpful solution for a laptop.
> > Mind that I come from a Ubuntu distro, Windows is the most energy efficient solution at the moment and allows me to run my Dev tools and work on battery for 4 hours straight. Ubuntu (or any Debian-based distro) destroy my battery in 40 minutes and there's no solution or optimization for that.
> 
> There is a solution for that: turn off your dedicated GPU, because that's the only possible way I can see that happening on Ubuntu. Going around crowing about the lack of a solution just shows how little you actually care about fixing it.

This is not quite the place for the discussion as someone pointed, I've gone great lengths to minimise battery consumption and yes, that included switching off my dedicated GPU or tweaking configs to no end.

My point was about remarking that "switch to a different os" is never a valid answer when someone points out a bug on a specific platform, both because one has reasons why it is using a specific operating system and because "use something else" doesn't solve the issue.
Would you go buy a new car if your mechanic said that he can't unlock the door to your car?

**Comment by mohamedhaddi** [comment]
Just out of curiosity, those having this issue which distro(s) are they using? I've been using Pengwin for around 6 months now and haven't had this particular problem (so far).

But I do remember a weird (and terrifying) problem that happened to me once (some weeks ago), not sure if it's related to this one or caused by WSL at all, but it's similar in the sense that it happened after putting my laptop to sleep, a file that I've been working on for hours, of which I'm 100% sure was saved, I put my laptop to sleep and come back an hour later to find that the file is completely empty. Of course, I made the mistake of not making any commits or pushing it to a remote (although it was in a git repo), but also I'm not *required* to do that to keep my files, it's totally not my mistake that the content of a file that I've saved countless times is lost just like that, because that never happened and it's just not supposed to happen. So what caused it exactly? I have no idea, but all I know is that I was working on Neovim on Pengwin on WSL2.

**Comment by okms** [comment]
I work from a Debian image in WSL2 with files living on the Linux side. I've not experienced this issue at all as far as I can remember. I run the `4.19.128-microsoft-standard` kernel.

**Comment by lindsm** [comment]
I don't have a lot of experience with WSL2, but just wanted to share mine in case it helps anybody.  Will be pretty random.

Recently, my work mac died from a dead battery.  While it was in for repair, I used a C940 Yoga I had.  I fresh installed Windows 10 on the Yoga, installed WSL2 & Ubuntu.  Set Ubuntu as my default shell in the new "Windows Terminal", so no powershell usage unless for a windows command.  I installed VS Code in windows, opened it from its shortcut, and used the settings sync.  Once I was satisfied Code was functioning normally under Windows (all addons installed, no more update prompts), I used Windows Terminal to git clone a popular work repo.  I opened code, in Windows Terminal, from that path using `code .`.  VS Code installed some extensions to deal with WSL at this time; `WSL Ubuntu-20.04...` with some Python and Jupyter stuff included.  Everything appeared to be working correctly.  I could `git clone`, run python, and run python poetry successfully.

I worked like this for about a month, would do my local development work, some minor python linting and testing, then push it up to a GitLab pipeline for full testing.  I used the VS Code terminal, set to WSL2 Ubuntu bash, like this for about a month.  Had many pushes to my teams repo with no issues.  It was a popular repo, so errors would of been noticed right away.  I know my laptop had gone to sleep during this, many times, with no issues.  The laptop was my gaming rig as well, with an eGPU attached.  So there was a lot of time with VS Code and Windows Terminal closed, never checked out the Windows Terminal got shut down, it just worked.

Overall, for me, a pleasant work experience compared to trying to run linux on the temporary laptop sent to me, while still waiting for my mac to be repaired.

**Comment by sangeeth96** [comment]
This used to happen to me and my co-workers on Mac/WSL when you're using something that periodically syncs a repo (like a `git fetch`) and the process gets terminated forcefully or the system shut-down abruptly. If you're using VSCode, you should probably disable this (should be called something like "Aurofetch") or if you are using some other program (maybe a feature of your shell?), might want to disable that as well. I had this happen when using WSL2 as well which ended up corrupting my got repo.

Otherwise, I have been using WSL2 with VSCode for a long while and haven't had any issues. Not sure if there's some new update that's causing this if it's not due to the ones I mentioned.

**Comment by bscott** [comment]
> This is a dupe of [microsoft/WSL#5026](https://github.com/microsoft/WSL/issues/5026), you can work around it by making sure not to use `shutdown /r /t 0` to shut the machine down. WSL isn't getting shut down properly on your computer and you're doing the equivalent of yanking the power on the device

I just don't shutdown that often ;)

**Comment by BurnyBoi** [comment]
> > Is there any direct solution for this, this happens once a day for me!
> 
> Use actual Linux such as Debian, Ubuntu or Pop OS. If necessary, you can run Windows inside VirtualBox. Or run Windows programs with Wine, Proton etc. And it doesn't corrupt your file system.

Thanks. Uninstalling Windows and installing Arch worked for me.

**Comment by eddebc** [comment]
> > > > > Is there any direct solution for this, this happens once a day for me!
> > > > 
> > > > 
> > > > Use actual Linux such as Debian, Ubuntu or Pop OS. If necessary, you can run Windows inside VirtualBox. Or run Windows programs with Wine, Proton etc. And it doesn't corrupt your file system.
> > > 
> > > 
> > > While I would do this for a tower/desktop that isn't an helpful solution for a laptop.
> > > Mind that I come from a Ubuntu distro, Windows is the most energy efficient solution at the moment and allows me to run my Dev tools and work on battery for 4 hours straight. Ubuntu (or any Debian-based distro) destroy my battery in 40 minutes and there's no solution or optimization for that.
> > 
> > 
> > There is a solution for that: turn off your dedicated GPU, because that's the only possible way I can see that happening on Ubuntu. Going around crowing about the lack of a solution just shows how little you actually care about fixing it.
> 
> This is not quite the place for the discussion as someone pointed, I've gone great lengths to minimise battery consumption and yes, that included switching off my dedicated GPU or tweaking configs to no end.
> 
> My point was about remarking that "switch to a different os" is never a valid answer when someone points out a bug on a specific platform, both because one has reasons why it is using a specific operating system and because "use something else" doesn't solve the issue.
> Would you go buy a new car if your mechanic said that he can't unlock the door to your car?

Not trying to be funny but that actually happened to me with an old ford. Was the last straw on a string of issues and ended up replacing it, yes. It came down to cost. Just like my workstation. It costs too much to debug issues with such a tightly closed source tool. It costs less on something like arch where it's possible for a community to come together to provide an actual fix. No one forbids it. And cost was the main driving factor for my change (that and naturally learning Linux this way)

**Comment by giannoug** [comment]
I've been having a similar issue with zsf where the history file sometimes gets corrupted. It seems that it happens when I don't close all WSL-related windows (VSCode and Windows Terminal) before shuting down the machine. My workaround is to close every VSCode window manually before shutting down the computer. I think that running "exit" on the terminal sessions helps instead of just closing the app. I've hadn't had this issue for weeks now 🥇

**Comment by dscho** [comment]
> 
> 
> Have you tried fiddling with [`core.fsyncObjectFiles`](https://git-scm.com/docs/git-config#Documentation/git-config.txt-corefsyncObjectFiles)?

I am just going to repeat this here because I fear it was drowned out.

The reason why I suggest this is that the `fsync()` call will make sure that the file contents will be written to disk. Otherwise, you will see files that have the correct size, but are filled with NUL bytes, when the filesystem driver did not have a chance to finalize those writes.

There have been some discussions revolving around enabling this flag by default in Git, but they petered out. Most likely because calling `fsync()` potentially hurts performance, and foregoing those calls will improve performance (but at the cost of risking lost data). See e.g. https://lore.kernel.org/git/20180117184828.31816-1-hch@lst.de/ (a thread from 2018 which has been revived late last year).

**Comment by luigimannoni** [comment]
Good point, will try as next thing however still won't fix corruption of bash history. I've also sustained data loss on a folder with binary/assets files but that might be correlated with git corruption

**Comment by craigloewen-msft** [comment]
We're looking into this issue! We're trying to reproduce it locally on our machines, so any repro steps to get into this state would be helpful (though it does seem tricky to get an exact repro), thank you to folks who are letting us know what happens when they get into this state. We'll post any updates here as they become available. Thank you!

**Comment by cmurf** [comment]
> The reason why I suggest this is that the `fsync()` call will make sure that the file contents will be written to disk. Otherwise, you will see files that have the correct size, but are filled with NUL bytes, when the filesystem driver did not have a chance to finalize those writes.

This is for crash safety, yes. But if this resolves the problem, it suggests the WSL2/VM is crashing. A proper shutdown means one of two things succeeds: umount or remount,ro. The former is typical with a separate /home file system. The latter is typical for sysroot. Either of those will flush everything to the file system including file system metadata.

Why is the VM being shutdown abruptly? Is the guest "hanging" on something that's stuck and can't remount,ro? And then the VM subsequently gets clobbered after some reasonable time out? Or is the VM being shutdown too soon?

**Comment by runvnc** [comment]
Just a random idea.  Any chance this is related to an issue with Fast Startup?

**Comment by KevinCathcart** [comment]
In case it helps, over on Hacker News there are a few interesting comments. One user has noted similar ext4 corruption in their normal Linux VMs when running on Hyper-V. This seems to suggest that there may be some form of general issue with Linux VMs. It is plausible that the use cases of WSL-2 make it more likely to impact these than general Hyper-V VMs. (For example it might use some vhdx options that are less commonly used by production linux VMs?) Or the noted problem might have an unrelated cause.

Several comments speculate that the cause might be in the block IO drivers, or Hyper-V storage subsystem, with them not providing certain synchronization guarantees that Linux is expecting. The other primary speculated cause over there is that the hypervisor is terminating the VM too early in some cases. However with Linux on raw hardware pulling the plug early does not typically cause filesystem corruption, so violating one of the block IO synchronization assumptions would still be a factor.

Improper system shutdown process might also implicate the WSL2's custom init executable. But again, even if WSL2 always stopped by way of "pulling the plug" out from under the utility VM, the filesystem should not sustain corruption. That really implicates either the Hyper-V block storage system (or any Hyper-V or WSL2 storage specific Linux kernel code), or that the kernel image Microsoft provides lacks some important bugfix from upstream.

**Comment by craigloewen-msft** [comment]
We've identified the error that has caused this issue and have put in a fix to add a sync() call before shutdown to make sure that all IO is completed. This fix is currently being verified internally and we're working on making it available to all of you! I'll post any updates here when it becomes available and how to get it. Thank you for the reports and help debugging!

**Comment by craigloewen-msft** [comment]
This fix is now available in the latest Insiders preview build 21292. Could any folks here who see this issue please install this build and let us know if it resolves the issue for you? Thanks!

**Comment by peterjuras** [comment]
Will this fix make it to stable releases as well? Or will we have to wait for the next big half year update (á la 21 04 / 21 09)?

**Comment by craigloewen-msft** [comment]
We are investigating bringing this fix to stable releases as well, and I'll post updates on that here as they become available.

**Comment by dltmurphy** [comment]
> This fix is now available in the latest Insiders preview build 21292. Could any folks here who see this issue please install this build and let us know if it resolves the issue for you? Thanks!

Seems to be fixed for me on the new build
